File: fortran/src/high_scores.f90

module high_scores
  implicit none
contains

function scores(score_list)
  integer, dimension(:), intent(IN) :: score_list
  integer, dimension(size(score_list)) :: scores
  scores = score_list
end function

integer function latest(score_list)
  integer, dimension(:), intent(IN) :: score_list

  latest = score_list(size(score_list))
end function

integer function personalBest(score_list)
  integer, dimension(:), intent(IN) :: score_list
  personalBest = maxval(score_list)
end function

function personalTopThree(score_list)
  integer, dimension(:), intent(IN) :: score_list
  integer, dimension(size(score_list)) :: scores
  integer, dimension(3) :: personalTopThree
  integer :: i, l, sz

  scores = score_list
  sz = size(scores)

  ! only after the top 3 so the performance doesn't matter too much
  do i = 1, min(3, sz)
    l = MAXLOC(scores, DIM=1)
    personalTopThree(i) = scores(l)
    scores(l) = -9999
  end do

  ! fill in the rest of the array
  do i = min(3, sz) + 1, 3
    personalTopThree(i) = 0
  end do

end function

end module

File: java/src/main/java/HighScores.java
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

class HighScores {
    private final List<Integer> highScores;

    public HighScores(final List<Integer> highScores) {
        this.highScores = highScores;
    }

    List<Integer> scores() {
        return highScores;
    }

    Integer latest() {
        return highScores.get(highScores.size() - 1);
    }

    Integer personalBest() {
        return highScores.stream().max(Comparator.naturalOrder()).orElse(0);
    }

    List<Integer> personalTopThree() {
        return highScores.stream().sorted(Comparator.reverseOrder()).limit(3).toList();
    }

}
